### PR TITLE
Add/fix error handling in some functions triggered by DSA.

### PR DESCRIPTION
**Pull request by PavelKopyl**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
We need you to sign our CLA, https://www.openssl.org/policies/cla.html, or to put in the commit message "CLA: trivial"  That means you think this is not copyrightable.  (And then the reviewers have to agree.)

Thanks for the fixes!

File: crypto/asn1/a_i2d_fp.c [hunk 1]
```diff
@@ -38,6 +38,9 @@ int ASN1_i2d_bio(i2d_of_void *i2d, BIO *out, unsigned char *x)
     int i, j = 0, n, ret = 1;
 
     n = i2d(x, NULL);
+    if (n < 0)
```
Comment (kroeckx) [inline crypto/asn1/a_i2d_fp.c]: That should probably be <= 0

**Comment by bernd-edlinger** [comment]
except the x_pubkey changes this should also apply to 1.0.2

**Comment by mattcaswell** [comment]
Ping @kroeckx

**Comment by bernd-edlinger** [comment]
merged to master: 7760384b4 a6f622bc9 157997f0c
1.1.0: bf37a9f58 64adaa3
1.0.2: 950d49d 200de40

Thanks!

**Comment by davidben** [comment]
Did you mean to import 157997f0c00f63b6005f3b5b6e0bb659d795e136 into 1.1.0 and 1.0.2 too? It looks like that also applies as well.

Although if the intent is that i2d functions return -1 for error and 0 for writing no bytes, at a quick glance `ASN1_item_ex_i2d` has its own incorrect return zeros (`asn1_enc_restore` and `asn1_cb`). And then `asn1_template_ex_i2d` checks `ASN1_item_ex_i2d`'s return value wrong.

**Comment by kroeckx** [comment]
I would have expected that to end up in all branches too.

The i2d_TYPE functions are documented to return a negative value on failure. ASN1_item_i2d and ASN1_item_ex_i2d are unfortunately not documented. But the i2d_TYPE functions are implemented by calling ASN1_item_i2d, so it at least applies to that function too. I see no reason why ASN1_item_ex_i2d shouldn't also return -1 on failure.

**Comment by kroeckx** [comment]
Since ASN1_item_i2d does `return asn1_item_flag_i2d()` and that in turn does `return ASN1_item_ex_i2d()`, it seems that ASN1_item_ex_i2d should also return a negative number for failures.

**Comment by bernd-edlinger** [comment]
yes, I thought that would work, but the cherry pick failed.
But if you consider that important, I could fix it up manually?

**Comment by kroeckx** [comment]
If uncertain, just open a new pull request.

**Pull request closed without merging**